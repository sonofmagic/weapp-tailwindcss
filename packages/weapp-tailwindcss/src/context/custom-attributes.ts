import type {
  ICustomAttributes,
  ICustomAttributesEntities,
  ItemOrItemArray,
} from '@/types'
import { isMap } from '@/utils'

export function toCustomAttributesEntities(customAttributes: ICustomAttributes): ICustomAttributesEntities {
  if (isMap(customAttributes)) {
    return [
      ...(customAttributes as Exclude<ICustomAttributes, Record<string, ItemOrItemArray<string | RegExp>>>).entries(),
    ]
  }

  return Object.entries(customAttributes)
}
