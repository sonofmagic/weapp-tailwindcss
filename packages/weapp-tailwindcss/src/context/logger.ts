import { logger } from '@weapp-tailwindcss/logger'

// https://www.npmjs.com/package/consola
const loggerLevelMap: Record<'info' | 'warn' | 'error' | 'silent', number> = {
  error: 0,
  warn: 1,
  info: 3,
  silent: -999,
}

export type LoggerLevel = keyof typeof loggerLevelMap

export function applyLoggerLevel(logLevel?: LoggerLevel): void {
  logger.level = loggerLevelMap[logLevel ?? 'info'] ?? loggerLevelMap.info
}
