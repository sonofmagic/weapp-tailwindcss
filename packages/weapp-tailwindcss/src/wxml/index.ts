export { replaceWxml } from './shared'
export { generateCode, templateReplacer, customTemplateHandler, createTemplateHandler } from './utils'
