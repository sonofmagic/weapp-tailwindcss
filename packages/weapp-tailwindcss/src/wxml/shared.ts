import type { ITemplateHandlerOptions } from '@/types'
import { SimpleMappingChars2String, escape } from '@/escape'

export function replaceWxml(
  original: string,
  options: ITemplateHandlerOptions = {
    keepEOL: false,
    escapeMap: SimpleMappingChars2String,
  },
) {
  const { keepEOL, escapeMap, mangleContext } = options
  let res = original
  if (!keepEOL) {
    res = res
      // 去除无用换行符和空格
      // 不能全去掉，头条小程序变量绑定，实现方式依赖空格，你说坑不坑？
      .replaceAll(/[\n\r]+/g, '')
  }
  if (mangleContext) {
    res = mangleContext.wxmlHandler(res)
  }

  res = escape(res, {
    map: escapeMap,
  })

  return res
}
