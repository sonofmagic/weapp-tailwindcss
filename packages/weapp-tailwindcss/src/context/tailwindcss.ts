import type { TailwindcssPatchOptions } from 'tailwindcss-patch'
import type { InternalUserDefinedOptions } from '@/types'
import { createRequire } from 'node:module'
import path from 'node:path'
import process from 'node:process'
import { fileURLToPath } from 'node:url'
import { logger } from '@weapp-tailwindcss/logger'
import { createTailwindcssPatcher } from '@/tailwindcss'
import { defuOverrideArray } from '@/utils'

const ENV_BASEDIR_KEYS = [
  'WEAPP_TAILWINDCSS_BASEDIR',
  'WEAPP_TAILWINDCSS_BASE_DIR',
  'TAILWINDCSS_BASEDIR',
  'TAILWINDCSS_BASE_DIR',
  'UNI_INPUT_DIR',
  'UNI_INPUT_ROOT',
  'UNI_CLI_ROOT',
  'UNI_APP_INPUT_DIR',
  'INIT_CWD',
  'PWD',
] as const

type EnvBasedirKey = typeof ENV_BASEDIR_KEYS[number]

const GENERIC_ENV_BASEDIR_KEYS = new Set<EnvBasedirKey>(['INIT_CWD', 'PWD'])

interface EnvBasedirResult {
  key: EnvBasedirKey
  value: string
}

function pickEnvBasedir(): EnvBasedirResult | undefined {
  for (const key of ENV_BASEDIR_KEYS) {
    const value = process.env[key]
    if (value && path.isAbsolute(value)) {
      return { key, value }
    }
  }
  return undefined
}

function pickPackageEnvBasedir(): string | undefined {
  const packageJsonPath = process.env.npm_package_json
  if (packageJsonPath) {
    const packageDir = path.dirname(packageJsonPath)
    if (packageDir && path.isAbsolute(packageDir)) {
      return packageDir
    }
  }
  const localPrefix = process.env.npm_config_local_prefix
  if (localPrefix && path.isAbsolute(localPrefix)) {
    return localPrefix
  }
  return undefined
}

function detectCallerBasedir(): string | undefined {
  const stack = new Error('resolveTailwindcssBasedir stack probe').stack
  if (!stack) {
    return undefined
  }
  if (process.env.WEAPP_TW_DEBUG_STACK === '1') {
    logger.debug('caller stack: %s', stack)
  }
  const lines = stack.split('\n')
  for (const line of lines) {
    const match = line.match(/\(([^)]+)\)/u) ?? line.match(/at\s+(\S.*)$/u)
    const location = match?.[1]
    if (!location) {
      continue
    }
    let filePath = location
    if (filePath.startsWith('file://')) {
      try {
        filePath = fileURLToPath(filePath)
      }
      catch {
        continue
      }
    }
    const [candidate] = filePath.split(':')
    const resolvedPath = path.isAbsolute(filePath) ? filePath : candidate
    if (!path.isAbsolute(resolvedPath)) {
      continue
    }
    if (resolvedPath.includes('node_modules') && resolvedPath.includes('weapp-tailwindcss')) {
      continue
    }
    try {
      return path.dirname(resolvedPath)
    }
    catch {
      continue
    }
  }
  return undefined
}

export function resolveTailwindcssBasedir(basedir?: string) {
  const envBasedirResult = pickEnvBasedir()
  const envBasedir = envBasedirResult?.value
  const envBasedirKey = envBasedirResult?.key
  const envBasedirIsGeneric = envBasedirKey ? GENERIC_ENV_BASEDIR_KEYS.has(envBasedirKey) : false
  const packageEnvBasedir = pickPackageEnvBasedir()
  const shouldDetectCaller = !envBasedir || envBasedirIsGeneric
  const callerBasedir = shouldDetectCaller ? detectCallerBasedir() : undefined
  const anchor = envBasedir ?? packageEnvBasedir ?? callerBasedir ?? process.cwd()
  if (process.env.WEAPP_TW_DEBUG_STACK === '1') {
    logger.debug('resolveTailwindcssBasedir anchor %O', {
      basedir,
      envBasedir,
      envBasedirKey,
      envBasedirIsGeneric,
      packageEnvBasedir,
      callerBasedir,
      npm_package_json: process.env.npm_package_json,
      cwd: process.cwd(),
      anchor,
    })
  }

  if (basedir && basedir.trim().length > 0) {
    if (path.isAbsolute(basedir)) {
      return path.normalize(basedir)
    }
    return path.resolve(anchor, basedir)
  }

  if (envBasedir && !envBasedirIsGeneric) {
    return path.normalize(envBasedir)
  }

  if (packageEnvBasedir) {
    return path.normalize(packageEnvBasedir)
  }

  if (callerBasedir) {
    const normalizedCaller = path.normalize(callerBasedir)
    const librarySegment = `${path.sep}weapp-tailwindcss${path.sep}`
    if (!normalizedCaller.includes(librarySegment)) {
      return normalizedCaller
    }
  }

  if (envBasedir) {
    return path.normalize(envBasedir)
  }

  const packageName = process.env.PNPM_PACKAGE_NAME
  if (packageName) {
    try {
      const anchorRequire = createRequire(path.join(anchor, '__resolve_tailwindcss_basedir__.cjs'))
      const packageJsonPath = anchorRequire.resolve(`${packageName}/package.json`)
      if (process.env.WEAPP_TW_DEBUG_STACK === '1') {
        logger.debug('package basedir resolved from PNPM_PACKAGE_NAME: %s', packageJsonPath)
      }
      return path.normalize(path.dirname(packageJsonPath))
    }
    catch {
      if (process.env.WEAPP_TW_DEBUG_STACK === '1') {
        logger.debug('failed to resolve package json for %s', packageName)
      }
      // ignore
    }
  }

  return path.normalize(process.cwd())
}

export function createTailwindcssPatcherFromContext(ctx: InternalUserDefinedOptions) {
  const {
    tailwindcssBasedir,
    supportCustomLengthUnitsPatch,
    tailwindcss,
    tailwindcssPatcherOptions,
    cssEntries,
    appType,
  } = ctx

  const resolvedTailwindcssBasedir = resolveTailwindcssBasedir(tailwindcssBasedir)
  ctx.tailwindcssBasedir = resolvedTailwindcssBasedir
  logger.debug('tailwindcss basedir resolved: %s', resolvedTailwindcssBasedir)

  type TailwindUserOptions = NonNullable<TailwindcssPatchOptions['tailwind']>

  const defaultTailwindcssConfig: TailwindUserOptions = {
    cwd: resolvedTailwindcssBasedir,
    v2: {
      cwd: resolvedTailwindcssBasedir,
    },
    v3: {
      cwd: resolvedTailwindcssBasedir,
    },
    v4: {
      base: resolvedTailwindcssBasedir,
      cssEntries,
    },
  }

  if (cssEntries?.length && (tailwindcss == null || tailwindcss.version == null)) {
    defaultTailwindcssConfig.version = 4
  }

  return createTailwindcssPatcher(
    {
      basedir: resolvedTailwindcssBasedir,
      cacheDir: appType === 'mpx' ? 'node_modules/tailwindcss-patch/.cache' : undefined,
      supportCustomLengthUnitsPatch: supportCustomLengthUnitsPatch ?? true,
      tailwindcss: defuOverrideArray<TailwindUserOptions, TailwindUserOptions[]>(
        (tailwindcss ?? {}) as TailwindUserOptions,
        defaultTailwindcssConfig,
      ),
      tailwindcssPatcherOptions,
    },
  )
}
