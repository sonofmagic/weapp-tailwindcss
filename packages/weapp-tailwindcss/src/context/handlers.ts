import type { IMangleScopeContext } from '@weapp-tailwindcss/mangle'
import type { ICustomAttributesEntities, InternalUserDefinedOptions } from '@/types'
import { createStyleHandler } from '@weapp-tailwindcss/postcss'
import { createJsHandler } from '@/js'
import { createTemplateHandler } from '@/wxml'

export function createHandlersFromContext(
  ctx: InternalUserDefinedOptions,
  mangleContext: IMangleScopeContext,
  customAttributesEntities: ICustomAttributesEntities,
  cssCalcOptions: InternalUserDefinedOptions['cssCalc'],
) {
  const {
    cssPreflight,
    customRuleCallback,
    cssPreflightRange,
    escapeMap,
    cssChildCombinatorReplaceValue,
    injectAdditionalCssVarScope,
    cssSelectorReplacement,
    rem2rpx,
    postcssOptions,
    cssRemoveProperty,
    cssRemoveHoverPseudoClass,
    cssPresetEnv,
    uniAppX,
    px2rpx,
    arbitraryValues,
    jsPreserveClass,
    babelParserOptions,
    ignoreCallExpressionIdentifiers,
    ignoreTaggedTemplateExpressionIdentifiers,
    inlineWxs,
    disabledDefaultTemplateHandler,
  } = ctx

  const styleHandler = createStyleHandler({
    cssPreflight,
    customRuleCallback,
    cssPreflightRange,
    escapeMap,
    mangleContext,
    cssChildCombinatorReplaceValue,
    injectAdditionalCssVarScope,
    cssSelectorReplacement,
    rem2rpx,
    postcssOptions,
    cssRemoveProperty,
    cssRemoveHoverPseudoClass,
    cssPresetEnv,
    uniAppX,
    cssCalc: cssCalcOptions,
    px2rpx,
  })

  const jsHandler = createJsHandler({
    escapeMap,
    mangleContext,
    arbitraryValues,
    jsPreserveClass,
    generateMap: true,
    babelParserOptions,
    ignoreCallExpressionIdentifiers,
    ignoreTaggedTemplateExpressionIdentifiers,
    uniAppX,
  })

  const templateHandler = createTemplateHandler({
    customAttributesEntities,
    escapeMap,
    mangleContext,
    inlineWxs,
    jsHandler,
    disabledDefaultTemplateHandler,
  })

  return {
    styleHandler,
    jsHandler,
    templateHandler,
  }
}
