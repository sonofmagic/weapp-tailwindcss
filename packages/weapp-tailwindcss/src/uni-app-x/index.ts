import type { AttributeNode, DirectiveNode, ParentNode } from '@vue/compiler-dom'
import type { TransformResult } from 'vite'
import type { CreateJsHandlerOptions, JsHandler } from '@/types'
import { NodeTypes } from '@vue/compiler-dom'
import { parse } from '@vue/compiler-sfc'
import MagicString from 'magic-string'
import { generateCode, replaceWxml } from '@/wxml'

function traverse(node: ParentNode, visitor: (node: ParentNode) => void): void {
  visitor(node)
  if (Array.isArray(node.children)) {
    for (const child of node.children) {
      if (child && typeof child === 'object' && 'type' in child) {
        traverse(child as ParentNode, visitor)
      }
    }
  }
}

function createClassAttributeUpdater(ms: MagicString) {
  return (prop: AttributeNode) => {
    if (prop.value) {
      ms.update(prop.value.loc.start.offset + 1, prop.value.loc.end.offset - 1, replaceWxml(prop.value.content))
    }
  }
}

function createClassDirectiveUpdater(ms: MagicString, jsHandler: JsHandler, runtimeSet?: Set<string>) {
  return (prop: DirectiveNode) => {
    if (prop.arg?.type === NodeTypes.SIMPLE_EXPRESSION && prop.arg.content === 'class' && prop.exp?.type === NodeTypes.SIMPLE_EXPRESSION) {
      const generated = generateCode(prop.exp.content, {
        jsHandler,
        runtimeSet,
      })
      ms.update(prop.exp.loc.start.offset, prop.exp.loc.end.offset, generated)
    }
  }
}

const defaultCreateJsHandlerOptions: CreateJsHandlerOptions = {
  babelParserOptions: {
    plugins: [
      'typescript',
    ],
  },
}
export function transformUVue(code: string, id: string, jsHandler: JsHandler, runtimeSet?: Set<string>): undefined | TransformResult {
  if (!/\.uvue(?:\?.*)?$/.test(id)) {
    return
  }
  const ms = new MagicString(code)
  const { descriptor, errors } = parse(code)
  if (errors.length === 0) {
    if (descriptor.template?.ast) {
      const updateStaticClass = createClassAttributeUpdater(ms)
      const updateDynamicClass = createClassDirectiveUpdater(ms, jsHandler, runtimeSet)
      traverse(descriptor.template.ast, (node) => {
        if (node.type !== NodeTypes.ELEMENT) {
          return
        }
        for (const prop of node.props) {
          if (prop.type === NodeTypes.ATTRIBUTE && prop.name === 'class') {
            updateStaticClass(prop)
          }
          else if (prop.type === NodeTypes.DIRECTIVE && prop.name === 'bind') {
            updateDynamicClass(prop)
          }
        }
      })
    }

    if (descriptor.script) {
      const { code } = jsHandler(descriptor.script.content, runtimeSet ?? new Set(), defaultCreateJsHandlerOptions)
      ms.update(descriptor.script.loc.start.offset, descriptor.script.loc.end.offset, code)
    }
    if (descriptor.scriptSetup) {
      const { code } = jsHandler(descriptor.scriptSetup.content, runtimeSet ?? new Set(), defaultCreateJsHandlerOptions)
      ms.update(descriptor.scriptSetup.loc.start.offset, descriptor.scriptSetup.loc.end.offset, code)
    }
  }
  return {
    code: ms.toString(),
    // @ts-ignore
    get map() {
      return ms.generateMap()
    },
  }
}
