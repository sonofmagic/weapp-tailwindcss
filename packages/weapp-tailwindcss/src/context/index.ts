import type { InternalUserDefinedOptions, UserDefinedOptions } from '@/types'
import { logger, pc } from '@weapp-tailwindcss/logger'
import { useMangleStore } from '@weapp-tailwindcss/mangle'
import { initializeCache } from '@/cache'
import { getDefaultOptions } from '@/defaults'
import { defuOverrideArray } from '@/utils'
import { toCustomAttributesEntities } from './custom-attributes'
import { createHandlersFromContext } from './handlers'
import { applyLoggerLevel } from './logger'
import { createTailwindcssPatcherFromContext } from './tailwindcss'

/**
 * 获取用户定义选项的内部表示，并初始化相关的处理程序和补丁。
 * @param opts - 用户定义的选项，可选。
 * @returns 返回一个包含内部用户定义选项的对象，包括样式、JS和模板处理程序，以及Tailwind CSS补丁。
 */
export function getCompilerContext(opts?: UserDefinedOptions): InternalUserDefinedOptions {
  const ctx = defuOverrideArray<InternalUserDefinedOptions, Partial<InternalUserDefinedOptions>[]>(
    opts as InternalUserDefinedOptions,
    getDefaultOptions() as InternalUserDefinedOptions,
    {},
  )

  ctx.escapeMap = ctx.customReplaceDictionary

  applyLoggerLevel(ctx.logLevel)

  const twPatcher = createTailwindcssPatcherFromContext(ctx)

  if (twPatcher.packageInfo?.version) {
    logger.success(`当前使用 ${pc.cyanBright('Tailwind CSS')} 版本为: ${pc.underline(pc.bold(pc.green(twPatcher.packageInfo.version)))}`)
  }
  else {
    logger.warn(`${pc.cyanBright('Tailwind CSS')} 未安装，已跳过版本检测与补丁应用。`)
  }

  const cssCalcOptions = ctx.cssCalc ?? twPatcher.majorVersion === 4

  const customAttributesEntities = toCustomAttributesEntities(ctx.customAttributes)

  const { initMangle, mangleContext, setMangleRuntimeSet } = useMangleStore()
  initMangle(ctx.mangle)

  const { styleHandler, jsHandler, templateHandler } = createHandlersFromContext(
    ctx,
    mangleContext,
    customAttributesEntities,
    cssCalcOptions,
  )

  ctx.styleHandler = styleHandler
  ctx.jsHandler = jsHandler
  ctx.templateHandler = templateHandler

  ctx.setMangleRuntimeSet = setMangleRuntimeSet
  ctx.cache = initializeCache(ctx.cache)
  ctx.twPatcher = twPatcher
  return ctx
}
