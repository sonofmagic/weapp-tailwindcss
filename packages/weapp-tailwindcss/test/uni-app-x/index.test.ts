import type { JsHandler } from '@/types'
import { createGetCase, fixturesRootPath } from '#test/util'
import { vi } from 'vitest'
import { getCompilerContext } from '@/context'
import { transformUVue } from '@/uni-app-x'
import { replaceWxml } from '@/wxml'

const getCase = createGetCase(fixturesRootPath)

describe('uni-app-x', () => {
  it('index.uvue', async () => {
    const { jsHandler } = getCompilerContext()
    const vueRawCode = await getCase('uni-app-x/index.uvue')
    const classNameSet = new Set<string>()
    classNameSet.add('text-[#258f27]')
    classNameSet.add('text-[100px]')
    classNameSet.add('py-[22.32px]')

    expect(transformUVue(vueRawCode, 'index.uvue', jsHandler, classNameSet)).toMatchSnapshot()
  })

  it('app.uvue', async () => {
    const { jsHandler } = getCompilerContext()
    const vueRawCode = await getCase('uni-app-x/App.uvue')
    const classNameSet = new Set<string>()
    classNameSet.add('text-[#258f27]')
    classNameSet.add('text-[100px]')
    classNameSet.add('py-[22.32px]')

    expect(transformUVue(vueRawCode, 'App.uvue', jsHandler, classNameSet)).toMatchSnapshot()
  })

  it('setup-lang-uts.uvue', async () => {
    const { jsHandler } = getCompilerContext()
    const vueRawCode = await getCase('uni-app-x/setup-lang-uts.uvue')
    const classNameSet = new Set<string>()
    classNameSet.add('text-[#258f27]')
    classNameSet.add('text-[100px]')
    classNameSet.add('py-[22.32px]')
    classNameSet.add('bg-[#000]')
    classNameSet.add('bg-[#111]')
    classNameSet.add('bg-[#222]')

    expect(transformUVue(vueRawCode, 'setup-lang-uts.uvue', jsHandler, classNameSet)).toMatchSnapshot()
  })

  it('transforms static and dynamic class bindings', () => {
    const runtimeSet = new Set<string>()
    const jsHandler: JsHandler = vi.fn((source: string) => ({ code: `handled(${source})` }))
    const code = `
<template>
  <view class="text-[#123]" :class="dynamicCls">
    <text>{{ label }}</text>
  </view>
</template>
<script lang="ts">
const label = 'hi'
</script>
`
    const result = transformUVue(code, 'sample.uvue', jsHandler, runtimeSet)
    expect(result?.code).toContain(`class="${replaceWxml('text-[#123]')}"`)
    expect(result?.code).toContain('handled(dynamicCls)')
    expect(result?.code).toContain('handled(\nconst label = \'hi\'\n)')
    expect(jsHandler).toHaveBeenCalled()
  })

  it('ignores non-uvue files', () => {
    const jsHandler: JsHandler = vi.fn((source: string) => ({ code: source }))
    const result = transformUVue('<template><view/></template>', 'App.vue', jsHandler, new Set())
    expect(result).toBeUndefined()
    expect(jsHandler).not.toHaveBeenCalled()
  })
})
