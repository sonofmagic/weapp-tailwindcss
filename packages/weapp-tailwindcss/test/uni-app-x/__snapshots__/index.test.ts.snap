// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`uni-app-x > app.uvue 1`] = `
{
  "code": "<script lang="uts">

	let firstBackTime = 0
	export default {
		onLaunch: function () {
			console.log('App Launch')
		},
		onShow: function () {
			console.log('App Show')
		},
		onHide: function () {
			console.log('App Hide')
		},

		onLastPageBackPress: function () {
			console.log('App LastPageBackPress')
			if (firstBackTime == 0) {
				uni.showToast({
					title: '再按一次退出应用',
					position: 'bottom',
				})
				firstBackTime = Date.now()
				setTimeout(() => {
					firstBackTime = 0
				}, 2000)
			} else if (Date.now() - firstBackTime < 2000) {
				firstBackTime = Date.now()
				uni.exit()
			}
		},

		onExit: function () {
			console.log('App Exit')
		},
	}
</script>

<style lang="scss">
@tailwind base;
@tailwind components;
@tailwind utilities;







</style>",
  "map": SourceMap {
    "file": undefined,
    "mappings": "AAAA,mBAAmB;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAmCnB;;AAEA;AACA;AACA;AACA;;;;;;;;AAQA",
    "names": [],
    "sources": [
      "",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;

exports[`uni-app-x > index.uvue 1`] = `
{
  "code": "<template>
	<view class="px-4">
		<view class="mt-_32d43rpx_ bg-_h322323_" :class="['text-_h258f27_', aaa]">
			<text class="text-_h258f27_">{{title}}</text>
		</view>
    <text class="icebreaker">icebreaker111111</text>
	</view>
</template>

<script>
	export default {
		data() {
			return {
				title: 'Hello',
        aaa: 'py-_22d32px_'
			}
		},
		onLoad() {

		},
		methods: {

		}
	}
</script>

<style>
.icebreaker{
  color: #844343;
}
</style>",
  "map": SourceMap {
    "file": undefined,
    "mappings": "AAAA;AACA,cAAc,IAAI;AAClB,eAAe,0BAA0B,UAAU,uBAAuB;AAC1E,gBAAgB,cAAc;AAC9B;AACA,iBAAiB,UAAU;AAC3B;AACA;;AAEA,QAAQ;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAeR;;AAEA;AACA;AACA;AACA;AACA",
    "names": [],
    "sources": [
      "",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;

exports[`uni-app-x > setup-lang-uts.uvue 1`] = `
{
  "code": "<script setup lang="uts">
	import WeappTailwindcss from '@/components/WeappTailwindcss.uvue'
	import { state, setGlobalNum } from '@/stores/counter'

	const buttonColors = [
		'bg-_h000_',
		'bg-_h111_',
		'bg-_h222_',
		'bg-[#333]',
		'bg-[#444]',
		'bg-[#555]',
		'bg-[#666]',
		'bg-[#777]',
		'bg-[#888]',
		'bg-[#999]',
		'bg-[#aaa]',
		'bg-[#bbb]',
		'bg-[#ccc]',
		'bg-[#ddd]',
		'bg-[#eee]',
		'bg-[#fff]',
	]
	const title = ref('Hello')
	const themeRef = ref<string>('')
	const classArray = computed(() => [
		'bg-[#aa00aa] text-[#ffffffee]',
	])
	const buttonClass = computed(() => {
		return buttonColors[state.globalNum % buttonColors.length]
	})






	onBeforeUnmount(() => {





	})

	onLoad(() => {
		console.log('欢迎使用 weapp-tailwindcss 模板')
	})

	function copy(data : string) {
		uni.setClipboardData({
			data,
		})
	}
</script>

<template>
	<view class="content">
		<WeappTailwindcss />
		<view class="my-3 w-full border-t border-solid border-gray-200" />
		<text class="text-xl text-gray-600s95">
			写法示例Start!
		</text>
		<view class="mt-_13d14758px_ flex flex-col items-center">
			<text class="w-32 rounded-md bg-pink-500 py-2 text-center font-semibold text-white ring-4 ring-pink-300">
				Default Ring
			</text>
			<view>
				<button class="w-64 text-white" :class="buttonClass" @click="setGlobalNum(state.globalNum+1)">
					click here to inc {{ state.globalNum }}
				</button>
			</view>
			<view class="test">
				<text class="text-_hffffff_ text-center">
					@apply 的用法，详见 \`index.vue\` 的 \`style\` 标签块
				</text>
			</view>
		</view>
	</view>
</template>

<style lang="scss" scoped>
	.content {
		@apply flex flex-col items-center py-4;
	}

	.test {
		@apply flex h-[100px] w-[222.222px] items-center justify-center rounded-[40px] bg-[#123456] bg-opacity-[0.54] #{!important};
	}
</style>",
  "map": SourceMap {
    "file": undefined,
    "mappings": "AAAA,yBAAyB;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAqDzB;;AAEA;AACA,cAAc,OAAO;AACrB;AACA,eAAe,iDAAiD;AAChE,eAAe,wBAAwB;AACvC;AACA;AACA,eAAe,0CAA0C;AACzD,gBAAgB,0FAA0F;AAC1G;AACA;AACA;AACA,mBAAmB,eAAe,UAAU,WAAW;AACvD;AACA;AACA;AACA,gBAAgB,IAAI;AACpB,iBAAiB,0BAA0B;AAC3C;AACA;AACA;AACA;AACA;AACA;;AAEA;AACA;AACA;AACA;;AAEA;AACA;AACA;AACA",
    "names": [],
    "sources": [
      "",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;
